File: frontend/src/lib/alerts.ts
import type { Alert, AlertKind } from "../types";
import { num, shortDate } from "./format";
import { type T, en } from "./i18n";

export type Severity = Alert["severity"];

const SEVERITY_RANK: Record<Severity, number> = { high: 0, medium: 1, info: 2 };
const KIND_ORDER: AlertKind[] = ["notice", "rough_sea", "busy", "long_wait", "rate_move", "rate_range", "weather_unavailable"];

export interface AlertGroup {
  key: string;
  kind: AlertKind;
  severity: Severity;
  title: string;
  /** One line under the title; lists the worst items first. */
  detail: string;
  items: Alert[];
}

function worst(items: Alert[]): Severity {
  return items.reduce<Severity>((s, a) => (SEVERITY_RANK[a.severity] < SEVERITY_RANK[s] ? a.severity : s), "info");
}

function where(a: Alert) {
  return a.port ?? a.vessel_class ?? "";
}

/** The short fact for one alert inside a group: "Paradip 4.1 m". */
function itemFact(a: Alert, t: T): string {
  switch (a.kind) {
    case "rough_sea":
      return `${where(a)} ${num(a.value ?? 0, 1)} m`;
    case "long_wait":
      return `${where(a)} ${num(a.wait_min_days ?? 0)}–${num(a.wait_max_days ?? 0)} d`;
    case "busy":
      return `${where(a)} +${num(a.value ?? 0)}%`;
    case "rate_move":
      return `${where(a)} ${a.direction === "down" ? "−" : "+"}${num(Math.abs(a.value ?? 0))}%`;
    case "rate_range":
      return `${where(a)} ±${num((a.value ?? 0) / 2)}%`;
    case "notice":
      return `${where(a)}: ${a.title}`;
    default:
      return t(where(a));
  }
}

/** The full sentence when a group has a single alert. */
function singleDetail(a: Alert, t: T): string {
  switch (a.kind) {
    case "rough_sea":
      return t("Waves up to {h} m on {date}; {over} of the next {days} days at or above {limit} m.", {
        h: num(a.value ?? 0, 1),
        date: shortDate(a.peak_date ?? a.date ?? ""),
        over: a.days_over ?? 0,
        days: a.days_total ?? 0,
        limit: num(a.threshold ?? 0, 1),
      });
    case "long_wait":
      return t("Ships typically wait {min}–{max} days for a berth.", { min: num(a.wait_min_days ?? 0), max: num(a.wait_max_days ?? 0) });
    case "busy":
      return t("{calls} dry-bulk calls a day over the last four weeks, {pct}% above the year before.", {
        calls: num(a.recent_calls_per_day ?? 0, 2),
        pct: num(a.value ?? 0),
      });
    case "rate_move":
      return t(a.direction === "down" ? "Forecast to fall {pct}% by the week of {date}." : "Forecast to rise {pct}% by the week of {date}.", {
        pct: num(Math.abs(a.value ?? 0)),
        date: shortDate(a.date ?? ""),
      });
    case "rate_range":
      return t("The likely range in {weeks} weeks spans {pct}% of today's rate.", { weeks: a.days_total ?? 0, pct: num(a.value ?? 0) });
    case "notice":
      return a.peak_date
        ? t("{title}, {from} to {to}.", { title: a.title, from: shortDate(a.date ?? ""), to: shortDate(a.peak_date) })
        : t("{title}, {from} onwards.", { title: a.title, from: shortDate(a.date ?? "") });
    case "weather_unavailable":
      return t("The marine weather service didn't respond; sea-state alerts will return when it does.");
    default:
      return a.message;
  }
}

function groupTitle(kind: AlertKind, items: Alert[], t: T): string {
  const n = items.length;
  const one = items[0];
  const port = t(where(one));
  switch (kind) {
    case "rough_sea":
      return n === 1 ? t("Rough sea at {port}", { port }) : t("Rough sea at {n} ports", { n });
    case "long_wait":
      return n === 1 ? t("Long waits at {port}", { port }) : t("Long waits at {n} ports", { n });
    case "busy":
      if (n > 1) return t("{n} ports busier than usual", { n });
      return one.scope === "load" ? t("Busier than usual at {port} (loading)", { port }) : t("Busier than usual at {port}", { port });
    case "notice":
      return n === 1 ? t("Notice at {port}", { port }) : t("Notices at {n} ports", { n });
    case "rate_move": {
      if (n === 1) return t(one.direction === "down" ? "{cls} rates expected to fall" : "{cls} rates expected to rise", { cls: port });
      return t("Rates expected to move");
    }
    case "rate_range":
      return n === 1 ? t("{cls} rates uncertain", { cls: port }) : t("Rates uncertain for {n} vessel types", { n });
    default:
      return t("Sea-state forecast unavailable");
  }
}

const SHOWN_FACTS = 2;

/** Alerts grouped by what they're about, worst first; one line each. */
export function groupAlerts(alerts: Alert[], t: T = en): AlertGroup[] {
  const byKind = new Map<AlertKind, Alert[]>();
  for (const a of alerts) byKind.set(a.kind, [...(byKind.get(a.kind) ?? []), a]);
  const groups: AlertGroup[] = [];
  for (const [kind, raw] of byKind) {
    const items = [...raw].sort((a, b) => SEVERITY_RANK[a.severity] - SEVERITY_RANK[b.severity] || Math.abs(b.value ?? 0) - Math.abs(a.value ?? 0));
    const facts = items.map((a) => itemFact(a, t));
    const more = facts.length - SHOWN_FACTS;
    const detail =
      items.length === 1
        ? singleDetail(items[0], t)
        : facts.slice(0, SHOWN_FACTS).join(" · ") + (more > 0 ? ` · ${t("+{n} more", { n: more })}` : "");
    groups.push({ key: kind, kind, severity: worst(items), title: groupTitle(kind, items, t), detail, items });
  }
  return groups.sort(
    (a, b) => SEVERITY_RANK[a.severity] - SEVERITY_RANK[b.severity] || KIND_ORDER.indexOf(a.kind) - KIND_ORDER.indexOf(b.kind)
  );
}

/** Full sentence for one alert, used when a group is expanded. */
export function alertSentence(a: Alert, t: T = en): string {
  return singleDetail(a, t);
}

/** Alerts that need attention (not notes), for counts. */
export function actionable(alerts: Alert[]) {
  return alerts.filter((a) => a.severity !== "info");
}

/** Short flags for one port's row in a table: "Rough sea 4.1 m". */
export function portFlags(alerts: Alert[], port: string, t: T = en): { label: string; severity: Severity }[] {
  return alerts
    .filter((a) => a.port === port && a.severity !== "info")
    .sort((a, b) => SEVERITY_RANK[a.severity] - SEVERITY_RANK[b.severity])
    .map((a) => ({
      severity: a.severity,
      label:
        a.kind === "rough_sea"
          ? t("Rough sea {h} m", { h: num(a.value ?? 0, 1) })
          : a.kind === "busy"
            ? t("Busy +{pct}%", { pct: num(a.value ?? 0) })
            : a.kind === "long_wait"
              ? t("Long wait")
              : a.title,
    }));
}

/** "Rough sea at 6 ports · 2 ports busier than usual" */
export function alertSummary(alerts: Alert[], t: T = en): string {
  return groupAlerts(actionable(alerts), t)
    .map((g) => g.title)
    .join(" · ");
}

File: src/alerts.py
"""Early warnings: sea state, port congestion and freight-market volatility.

Each alert states what triggered it and the threshold it crossed. Thresholds
are parameters, so a desk can tighten or loosen them. Alongside the English
title/message, every alert carries structured fields (kind, value,
threshold, unit, date, days_over, peak_date, direction) so a client can
group alerts and write its own text, e.g. in another language.
"""
from dataclasses import dataclass

import numpy as np
import pandas as pd

from src.cost_engine import expected_wait_days


@dataclass
class AlertThresholds:
    wave_height_m: float = 2.5
    wait_days: float = 5.0
    activity_surge_pct: float = 30.0
    forecast_band_pct: float = 40.0
    forecast_move_pct: float = 10.0
    activity_recent_days: int = 28


def port_activity(daily_df: pd.DataFrame | None, recent_days: int = 28, baseline_days: int = 365) -> dict | None:
    """Recent dry-bulk calls per day against the year before that window. The
    all-time average would flag steady growth since 2019 as a surge."""
    if daily_df is None or daily_df.empty:
        return None
    df = daily_df.sort_values("date")
    as_of = df["date"].max()
    recent_start = as_of - pd.Timedelta(days=recent_days - 1)
    base_start = recent_start - pd.Timedelta(days=baseline_days)
    recent = df[df["date"] >= recent_start]["portcalls_dry_bulk"]
    base = df[(df["date"] >= base_start) & (df["date"] < recent_start)]["portcalls_dry_bulk"]
    if recent.empty or base.empty or base.mean() == 0:
        return None
    return {
        "as_of": as_of.strftime("%Y-%m-%d"),
        "recent_calls_per_day": round(float(recent.mean()), 2),
        "baseline_calls_per_day": round(float(base.mean()), 2),
        "pct_vs_normal": round((float(recent.mean()) - float(base.mean())) / float(base.mean()) * 100, 1),
    }


def _day(iso: str) -> str:
    d = pd.Timestamp(iso)
    return f"{d.day} {d.strftime('%b')}"


def _alert(**kwargs) -> dict:
    base = {"port": None, "vessel_class": None, "date": None, "unit": None, "kind": None, "days_over": None,
            "days_total": None, "peak_date": None, "direction": None, "scope": "discharge"}
    base.update(kwargs)
    return base


def weather_alerts(weather: dict, t: AlertThresholds, ports: list[str] | None = None) -> list[dict]:
    alerts = []
    for port, w in weather.items():
        if ports and port not in ports:
            continue
        heights = w.get("wave_height_max_m") if isinstance(w, dict) else None
        dates = w.get("dates") if isinstance(w, dict) else None
        if not heights or not dates or w.get("error"):
            continue
        over = [(i, h) for i, h in enumerate(heights) if h is not None and h >= t.wave_height_m]
        if not over:
            continue
        first_i, _ = over[0]
        peak_i, peak = max(over, key=lambda x: x[1])
        severe = peak >= 1.5 * t.wave_height_m or first_i <= 1
        alerts.append(
            _alert(
                id=f"weather-{port}",
                category="weather",
                kind="rough_sea",
                days_over=len(over),
                days_total=len(heights),
                peak_date=dates[peak_i],
                severity="high" if severe else "medium",
                port=port,
                title=f"Rough sea at {port}",
                message=f"Waves up to {peak:.1f} m on {_day(dates[peak_i])}; {len(over)} of the next {len(heights)} days at or above {t.wave_height_m:g} m.",
                value=round(float(peak), 1),
                threshold=t.wave_height_m,
                unit="m",
                date=dates[first_i],
            )
        )
    return alerts


def congestion_alerts(
    ports_df: pd.DataFrame, activity: dict[str, dict | None], t: AlertThresholds, ports: list[str] | None = None
) -> list[dict]:
    alerts = []
    for _, port in ports_df.iterrows():
        name = port["name"]
        if ports and name not in ports:
            continue
        wait = expected_wait_days(port)
        if wait >= t.wait_days:
            alerts.append(
                _alert(
                    id=f"wait-{name}",
                    category="congestion",
                    kind="long_wait",
                    wait_min_days=float(port["avg_wait_days_min"]),
                    wait_max_days=float(port["avg_wait_days_max"]),
                    severity="high" if wait >= 1.5 * t.wait_days else "medium",
                    port=name,
                    title=f"Long waits at {name}",
                    message=f"Vessels typically wait {port['avg_wait_days_min']:g}–{port['avg_wait_days_max']:g} days for a berth.",
                    value=round(wait, 1),
                    threshold=t.wait_days,
                    unit="days",
                )
            )
        act = activity.get(name)
        if act and act["pct_vs_normal"] >= t.activity_surge_pct:
            alerts.append(
                _alert(
                    id=f"activity-{name}",
                    category="congestion",
                    kind="busy",
                    recent_calls_per_day=act["recent_calls_per_day"],
                    baseline_calls_per_day=act["baseline_calls_per_day"],
                    severity="high" if act["pct_vs_normal"] >= 1.5 * t.activity_surge_pct else "medium",
                    port=name,
                    title=f"Busier than usual at {name}",
                    message=(
                        f"{act['recent_calls_per_day']:.2f} dry-bulk calls a day over the last {t.activity_recent_days} days, "
                        f"{act['pct_vs_normal']:.0f}% above the previous year's {act['baseline_calls_per_day']:.2f}."
                    ),
                    value=act["pct_vs_normal"],
                    threshold=t.activity_surge_pct,
                    unit="%",
                    date=act["as_of"],
                )
            )
    return alerts


def origin_congestion_alerts(origin_activity: dict[str, dict | None], t: AlertThresholds) -> list[dict]:
    """Loading ports busier than usual: queues there delay the voyage before it starts."""
    alerts = []
    for name, act in origin_activity.items():
        if act and act["pct_vs_normal"] >= t.activity_surge_pct:
            alerts.append(
                _alert(
                    id=f"activity-load-{name}",
                    category="congestion",
                    kind="busy",
                    scope="load",
                    severity="high" if act["pct_vs_normal"] >= 1.5 * t.activity_surge_pct else "medium",
                    port=name,
                    title=f"Busier than usual at {name} (loading)",
                    message=(
                        f"{act['recent_calls_per_day']:.2f} dry-bulk calls a day over the last {t.activity_recent_days} days, "
                        f"{act['pct_vs_normal']:.0f}% above the previous year's {act['baseline_calls_per_day']:.2f}."
                    ),
                    value=act["pct_vs_normal"],
                    threshold=t.activity_surge_pct,
                    unit="%",
                    date=act["as_of"],
                    recent_calls_per_day=act["recent_calls_per_day"],
                    baseline_calls_per_day=act["baseline_calls_per_day"],
                )
            )
    return alerts


def market_alerts(forecasts: dict[str, dict], t: AlertThresholds) -> list[dict]:
    """forecasts: {vessel_class: {current_rate, point, lower, upper, dates}}"""
    alerts = []
    for cls, f in forecasts.items():
        current = float(f["current_rate"])
        point = np.asarray(f["point"], dtype=float)
        lower = np.maximum(np.asarray(f["lower"], dtype=float), 0)
        upper = np.asarray(f["upper"], dtype=float)
        weeks = len(point)
        band = (upper[-1] - lower[-1]) / current * 100
        if band >= t.forecast_band_pct:
            alerts.append(
                _alert(
                    id=f"band-{cls}",
                    category="market",
                    kind="rate_range",
                    days_total=weeks,
                    severity="high" if band >= 1.5 * t.forecast_band_pct else "medium",
                    vessel_class=cls,
                    title=f"{cls} rates uncertain",
                    message=f"The likely range in {weeks} weeks spans {band:.0f}% of today's rate.",
                    value=round(float(band), 1),
                    threshold=t.forecast_band_pct,
                    unit="%",
                    date=f["dates"][-1],
                )
            )
        moves = (point - current) / current * 100
        i = int(np.argmax(np.abs(moves)))
        if abs(moves[i]) >= t.forecast_move_pct:
            direction = "rise" if moves[i] > 0 else "fall"
            alerts.append(
                _alert(
                    id=f"move-{cls}",
                    category="market",
                    kind="rate_move",
                    direction="up" if moves[i] > 0 else "down",
                    severity="high" if abs(moves[i]) >= 1.5 * t.forecast_move_pct else "medium",
                    vessel_class=cls,
                    title=f"{cls} rates expected to {direction}",
                    message=f"Forecast to {direction} {abs(moves[i]):.0f}% by the week of {_day(f['dates'][i])}.",
                    value=round(float(moves[i]), 1),
                    threshold=t.forecast_move_pct,
                    unit="%",
                    date=f["dates"][i],
                )
            )
    return alerts


def notice_alerts(notices: pd.DataFrame, today: pd.Timestamp, ports: list[str] | None = None, lookahead_days: int = 14) -> list[dict]:
    """Hand-kept disruption notices in force today or starting within two weeks."""
    alerts = []
    horizon = today + pd.Timedelta(days=lookahead_days)
    for i, n in notices.iterrows():
        if ports and n["port"] not in ports:
            continue
        start, end = n["start_date"], n["end_date"] if pd.notna(n["end_date"]) else pd.Timestamp.max
        if start > horizon or end < today:
            continue
        alerts.append(
            _alert(
                id=f"notice-{i}-{n['port']}",
                category="disruption",
                kind="notice",
                severity=n["severity"] if n["severity"] in ("high", "medium", "info") else "medium",
                port=n["port"],
                title=str(n["title"]),
                message=str(n["title"]),
                value=None,
                threshold=None,
                date=start.strftime("%Y-%m-%d"),
                peak_date=None if end == pd.Timestamp.max else end.strftime("%Y-%m-%d"),
            )
        )
    return alerts


SEVERITY_ORDER = {"high": 0, "medium": 1, "info": 2}


def sort_alerts(alerts: list[dict]) -> list[dict]:
    return sorted(alerts, key=lambda a: (SEVERITY_ORDER.get(a["severity"], 3), a["category"], a["id"]))
